### PR TITLE
[WIP] Add basic InstPackages.sh

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This adds to the repository a cleaned up copy of `InstPackages.sh`, the script which tries to build GAP packages. While this isn't perfect, it is I believe better than what we have had before (go and find the script). Of course, on almost all computers you won't be able to build every package (has anyone ever had every package built at once?)

There are many ways this could be improved -- the major obvious issue is that at present we assume that we are in the 'pkg' directory of GAP's source tree.

The main simplification is that I try to build as many packages as possible with one simple(ish) rule. That rule is:

* If there is a `configure` file, run `./configure && make`

32-bit builds on 64-bit OSes add one wrinkle: If the configure script was generated by `autoconf`, then add `CFLAGS=-m32 LDFLAGS=-m32 LOPTS=-m32 CXXFLAGS=-m32 && make`.

This lets us build the vast majority of packages. A few packages remain, most of which will be easy to fix (if updates are released).

There are a few issues I hit, just for people's information:

* It's not generally safe to run 'make clean' or 'make distclean', as this will then force us to rebuild documentation which seems to fail for some packages.

* It's also not safe to run makefiles where there is no configure, as that also often seems to end up failing for documentation related reasons.

### CODE REVIEW

**Comment by alex-konovalov** [comment]
Thanks.  Making the `InstPackages.sh` script shorter is certainly useful for the ease of its maintenance. I think it should not be merged in the present form however. This PR reintroduces `pkg` directory again, and this will cause problems e.g. when there is a symlink to `pkg` from the GAP release. We tried hard to get rid of `pkg` directory in the repository in the past. So the script like that should live in the GAP root directory.

My second concern was that the core GAP distribution could be wrapped with different selection of packages, so I am not sure that it's good to make such script a part of the core GAP system. However, it looks very generic, and for all new packages that use a standard build procedure it will work automatically without even a need to edit the script, so this may no longer be a concern. Even more, the step of building packages is being overlooked so often, and now we may be able to call it automatically. 

In summary, I am happy to merge this after it will be adjusted to not to create `pkg` directory.

**Comment by fingolfin** [comment]
Overall, nice work!

Very minor nitpick: You seem to mix tabs and spaces, I find that midly irritating. Perhaps we should agree one some indention guidelines? e.g. I prefer not using TABs at all, simply because everybody has different settings for them.

One wish: In my modified version of this script, I allow the user to optionally pass one argument to the script: The GAPROOT, i.e. the location where GAP is installed; the default value being "..". This would be useful for me, because I keep my package distribution separate from my GAP directory, so that I can use it with multiple versions of GAP.

**Comment by alex-konovalov** [comment]
Fiddling with `pkg` was likely the reason why Travis failed. But something strange going there - it did not fail testinstall.g. We have to check if error detection works and use new functionality where the failure detection will be based on exit code and not on grepping the output.

**Comment by alex-konovalov** [comment]
> 
> There are a few issues I hit, just for people's information:
> 
> 	• It's not generally safe to run 'make clean' or 'make distclean', as this will then force us to rebuild documentation which seems to fail for some packages.
> 
> 	• It's also not safe to run makefiles where there is no configure, as that also often seems to end up failing for documentation related reasons.


Could you please provide more details - it's not clear for me how documentation is involved here. Thanks!

**Comment by ChrisJefferson** [comment]
I'll clean up the tabs+spaces, and dangling spaces. My text editor has a nice "do all this stuff neatly" option, but I have to keep it turned off or it chews through all the GAP code whenever I open a file and edit a single line :\

The 'building documentation' issue is most easily provided with an example. If I go into 4ti2Interface and type `make clean && make`, then I get a failure building documentation (`Error, Did not find chunk Groebner1`). This kind of error occurs in several packages, not sure why, but it means I can't safely just run `make clean` in packages before building them.

**Comment by alex-konovalov** [comment]
@ChrisJefferson ah, so you had in mind building documentation _for packages_ and running `make clean` _for packages_.  Because this is a GAP tracker, I was confused assuming that this is somehow connected with the core system.

`grep`ping for `Groebner1` in 4ti2 4ti2Interface shows only two lines:
```
doc/Chapter_4ti2_functions.xml:47:<#Include Label="Groebner1">
gap/4ti2Interface.gd:65:#! <#Include Label="Groebner1">
```
so there is include without matching `<#GAPDoc Label="Groebner1"> ... <#/GAPDoc>` chunk. That's a problem in the package - could be that the release contains outdated manual or the manual must be built in a different way - @sebasguts, could you please have a look?

**Comment by ChrisJefferson** [comment]
Sorry, yes.

Actually, I didn't really want to build documentation for packages, I more just wants to 'make clean', because if (for example) you build packages 32-bit, then try to build them 64-bit, it fails with a 'make clean' inbetween.

**Comment by alex-konovalov** [comment]
@ChrisJefferson yes, that's why I think its maybe either an error in the package (too radical `make clean`) or the user is supposed to call `make clean` but not `make distclean` to preserve the documentation, and there is some issue with its manual. Let's see what @sebasguts thinks.

What about creating another script, `CleanPackages.sh`, following the same structure? Btw, I suggest to rename `InstPackages.sh` to something else, to underscore that it's different from its predecessor that may be still lying around (and don't keep it in `pkg` as I wrote above).

**Comment by ChrisJefferson** [comment]
Some more basic modifications -- script is now at bin/BuildPackages.sh. Still has to be run in the 'pkg' directory, tells users to move there if run from 'bin' (I could just move there of course, but it gets tricky when also supporting non-standard pkg directories)

Still lots of little ways to improve (occurs to me to only bother with directories with a `PackageInfo.g`, in case we are run in a stupid place).

**Comment by sebasguts** [comment]
So, I looked at 4ti2Interface, and the reason why this happens is the one @alex-konovalov already mentioned: The included chunk is not there, since the examples aren't packed in the release tarball. Thanks for finding this error. So this then solves the problem in 4ti2Interface.

Generally, I do not think that `make clean` is a good option. At least for homalg, `make clean` always deletes the documentation. It is always possible that we do not want to ship all files from which the documentation is created in the tarball, because maybe this files only contain partial examples which have use for the user or even might be confusing. However, I do not think we do this at the moment. `make distclean` seems to be the more sensible solution here. Also, rebuilding the documentation is not a good idea if it is shipped with the package, since it might diverge (for any reason) from the documentation which is in the package.

On the other hand, if someone (including me) ships a makefile in the tarball, all the targets should work. This means `make clean && make doc` should work, if it is there. This is by the way not the case in any homalg package ;).

**Comment by sebasguts** [comment]
One more thing to mention here: If the documentation is build using AutoDoc, then this rebuild would need AutoDoc. Since AutoDoc is not needed to load the package itself, I try to only list it as a suggested package. But if you rebuild the documentation, AutoDoc would be a needed package.

**Comment by ChrisJefferson** [comment]
I'll hold off on 'make clean' for now. In the longer term, we might want to give a set of requirements for package authors on what various targets should do.

The target I really need is "Please clean out any code linking to the GAP kernel", as there are several packages at the moment where after either switching 64-bit to 32-bit, or just upgrading the GAP kernel, do not rebuild kernel packages which leads to either failures to load package, or crashes.

However, this problem has always existed, so we don't have to fix it today.

**Comment by alex-konovalov** [comment]
So maybe it's not a good idea to run `./configure ; make` in each directory where this is runnable in the 1st instance. The current `InstPackages.sh` script explicitly lists packages that require compilation to build binaries. It does nothing for other packages that may use their own automation for various purposes in various ways.

If we would be standardising this, I would suggest that there should be a separate target to clean documentation, and `make clean` should never delete it.

**Comment by ChrisJefferson** [comment]
I would really like to move away from hard-wiring how every package is built -- if there are packages where we shouldn't be running `./configure; make` I'm happy to add exceptions for them. Do you know if there are any packages we shouldn't be trying to build?

**Comment by alex-konovalov** [comment]
@ChrisJefferson No, I only know which packages we should be trying to build, and I am completely agnostic regarding what other packages having Makefile are doing...

**Comment by fingolfin** [comment]
For the records, this is what my `InstPackages.sh` does:
```
#!/bin/sh -e

if [ $# -lt 1 ]; then
    GAPROOT=../..
else
    GAPROOT=$1
fi

for pkg in ace Browse edim example Gauss grape guava* ; do
    echo "==== Building $pkg ===="
    ( cd $pkg && ./configure $GAPROOT && make )
    echo ""
done

for pkg in anupq* cvec* io* orb* float* nq* json*; do
    echo "==== Building $pkg ===="
    ( cd $pkg && ./configure --with-gaproot=$GAPROOT && make )
    echo ""
done
```

First off, note the "-e" in the first line, which ensures that it exits upon the first error. Secondly, note that it supports compiling outside of the `pkg` dir (that's important for me). Lastly, it uses a hardcoded list of package names. That makes it slightly less flexible, but on the other hand, it ensures we only build packages for which we really know how to build them. As a consequence, it is also much simpler (but also less flexible).

That said, I clearly prefer the more generic approach Chris has taken, and I think adding GAPROOT support to it is fairly easy to achieve in the obvious way :).

As for "make clean": I would simply avoid using it for now.

**Comment by fingolfin** [comment]
Regarding the fact that some packages require `AutoDoc` in order to compiler their documentation: There are multiple ways to approach this. For example, we could add a new `BuildDepends` field to `PackageInfo.g`. This might be a good idea on the long run anyway, as it might also be useful for a proper "GAP package manager", should we ever get around to doing on (in a sense, these things are small steps towards having such a thing).

**Comment by fingolfin** [comment]
Oh, and finally, of course we also need a proper solution for building multiple variants of a package: 32 vs 64 bits, optimized vs. debug builds, possibly more. So the desire to "make clean" here is understandable, but in the end just one aspect of a bigger problem (just saying -- but of course this should not stop is from fixing small parts of the big problem right away, if we can)

**Comment by sebasguts** [comment]
@ChrisJefferson Would it really be a good idea to opt out packages? I think packages that do not have any code that needs to be compiled should be left untouched. Maybe a list of packages that need compilation is the better option. As far as it comes to them, they should be standartized, i.e. `make clean; ./configure; make` should work in them.

**Comment by alex-konovalov** [comment]
A mix of both approaches is needed, otherwise we are trying to build a framework for packages complying to some rules and automatically adding to it those which are not. I suggest to:
* have a list of packages that have Makefiles but don't have binaries, and specify in BuildPackages.sh that is does nothing for them
* Publicise rules for GAP packages to be compatible with BuildPackages.sh
* When new package appears and it complies with the rules, we do nothing
* If one of the excluded packages is updated and becomes compliant, we update the script.

**Comment by ChrisJefferson** [comment]
@sebasguts : I would really prefer not to give a list of packages which do need compiling, as such a list would become out-of-date very quickly, whereas (hopefully!) we can refuse to accept packages in future which have ./configure / Makefiles we should not be running.

**Comment by markuspf** [comment]
I think we should merge this and continue fix it on the go.

Rationale: This does not cause any disruptions to standard procedures, but improves the situation for people like me who install GAP in random places every so often.

**Comment by alex-konovalov** [comment]
I support merging this too. The only things that I think should be changed (can be done as post-merge cleanup) is that packages that currently do not support `./configure && make ;` should be explicitly excluded. For example, packages without binaries, where `make` is used to build manuals. We should then publicise specification for package authors standardising targets:
* default `make` should be called only for packages with binaries
* `make manuals` should be used to build manuals
* other targets - up to package authors.
* `make clean` cleans binaries if exist, but not manuals
* `make distclean` may clean all derived files, including manuals

**Comment by markuspf** [comment]
Of course I already have a nitpick, since on my system it is `gmake` not `make`, but I can fix this up and provide a pull request. I'll merge this now.

**Comment by markuspf** [comment]
Note that I edited the title to reflect the actual name of the script.
